pkg/agent/multicast/multicast_route.go

```diff
@@ -0,0 +1,414 @@
+
+	multicast_syscall "antrea.io/antrea/third_party/unix"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+const (
+	GroupNameIndexName = "groupName"
+	MulticastFlag      = "multicast"
+
+	igmpMsgminRetryDelay = 5 * time.Second
+	igmpMsgmaxRetryDelay = 300 * time.Second
+)
+
+func newRouteClient(nodeconfig *config.NodeConfig, ifaceStore interfacestore.InterfaceStore) (*MRouteClient, error) {
+	fd, err := createMulticastSocket()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create multicast socket")
+	}
+	VIFAllocator := newVIFAllocator(fd)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create multicast VIF allocator")
+	}
+	c := &MRouteClient{
+		sockFD:                 fd,
+		vifAllocator:           VIFAllocator,
+		gatewayInterface:       nodeconfig.GatewayConfig.Name,
+		ifaceStore:             ifaceStore,
+		nodeTransportInterface: nodeconfig.NodeTransportName,
+		inboundRouteCache:      cache.NewIndexer(getMulticastInboundEntryKey, cache.Indexers{GroupNameIndexName: inboundGroupIndexFunc}),
+		outboundRouteCache:     cache.NewIndexer(getMulticastOutboundEntryKey, cache.Indexers{GroupNameIndexName: outboundGroupIndexFunc}),
+		queue:                  workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(igmpMsgminRetryDelay, igmpMsgmaxRetryDelay), "multicastroute"),
+	}
+	c.externalAddrMap = c.getExternalMulticastAddrMap(nodeconfig.ExternalMulticastInterfaceConfigs, nodeconfig.NodeTransportIPv4Addr)
+	_, err = VIFAllocator.setVifIfNotExist(c.gatewayInterface)
+	if err != nil {
+		return nil, fmt.Errorf("failed to allocate VIF for gateway")
+	}
+	_, err = VIFAllocator.setVifIfNotExist(c.nodeTransportInterface)
+	if err != nil {
+		return nil, fmt.Errorf("failed to allocate VIF for transportInterface")
+	}
+
+	return c, nil
+}
+
+// MRouteClient configures static multicast route
+type MRouteClient struct {
+	sockFD                 int
+	vifAllocator           *VIFAllocator
+	gatewayInterface       string
+	nodeTransportInterface string
+	inboundRouteCache      cache.Indexer
+	outboundRouteCache     cache.Indexer
+	ifaceStore             interfacestore.InterfaceStore
+	externalAddrMap        []config.ExternalMulticastInterfaceConfig
+	queue                  workqueue.RateLimitingInterface
+}
+
+func createMulticastSocket() (int, error) {
+	fd, err := syscall.Socket(syscall.AF_INET, syscall.SOCK_RAW, syscall.IPPROTO_IGMP)
+	if err != nil {
+		return -1, fmt.Errorf("cannot create multicast socket")
+	}
+
+	err = syscall.SetsockoptInt(fd, syscall.IPPROTO_IP, multicast_syscall.MRT_BASE, 1)
+	if err != nil {
+		return -1, fmt.Errorf("unable to activte Multicast routing in kernel with error %s", err.Error())
+	}
+	return fd, nil
+}
+
+// externalInterfacesJoinMgroup allows external interfaces join multicast group,
+// making these interfaces accept mulitcast traffic with multicast ip:mgroup
```

Comment:
Add "." at end of lines. Please check other comments too.

Reply:
Updated to all the comments with periods.